staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go

```diff
@@ -0,0 +1,250 @@
+func RootCelContext(schema *apiextensions.JSONSchemaProps, converter Converter) CelSchemaContext {
+	rootCardinality := uint64(1)
+	r := CelSchemaContext{
+		MaxCardinality: &rootCardinality,
+		TotalCost:      &TotalCost{},
+		Converter:      converter,
+	}
+	r.initSchemaFields(schema, true)
+	return r
+}
+
+// ChildFieldContext constructs and returns a CelSchemaContext for fieldName.
+func (c *CelSchemaContext) ChildFieldContext(schema, propSchema *apiextensions.JSONSchemaProps, fieldName string) CelSchemaContext {
+	r := c.childContext(schema)
+	if c.Schema != nil && c.Schema.Properties != nil {
+		if propSchema, ok := c.Schema.Properties[fieldName]; ok {
+			r.Schema = &propSchema
+		}
+		if declField, ok := c.DeclType.Fields[fieldName]; ok {
+			r.DeclType = declField.Type
+		}
+	} else {
+		r.initSchemaFields(propSchema, false)
+	}
+	return r
+}
+
+// ChildValuesContext constructs and returns a CelSchemaContext for the properties of an object if this CelSchemaContext is an object.
+func (c *CelSchemaContext) ChildValuesContext(schema *apiextensions.JSONSchemaProps) CelSchemaContext {
+	r := c.childContext(schema)
+	if c.Schema != nil && c.Schema.AdditionalProperties != nil {
+		if itemsSchema := c.Schema.AdditionalProperties; itemsSchema != nil {
+			r.Schema = itemsSchema.Structural
+		}
+		r.DeclType = c.DeclType.ElemType
+	} else {
+		r.initSchemaFields(schema.AdditionalProperties.Schema, false)
+	}
+	return r
+}
+
+// ChildItemsContext constructs and returns a CelSchemaContext for the items of an array if this CelSchemaContext is an array.
+func (c *CelSchemaContext) ChildItemsContext(schema, itemsSchema *apiextensions.JSONSchemaProps) CelSchemaContext {
+	r := c.childContext(schema)
+	if c.Schema != nil && c.Schema.Items != nil {
+		if itemsSchema := c.Schema.Items; itemsSchema != nil {
+			r.Schema = itemsSchema
+		}
+		r.DeclType = c.DeclType.ElemType
+	} else {
+		r.initSchemaFields(itemsSchema, false)
+	}
+	return r
+}
+
+// ChildSchemaContext returns a CelSchemaContext for all child nodes that are not handled by ChildFieldContext,
+// ChildValuesContext or ChildItemsContext.
+func (c *CelSchemaContext) ChildSchemaContext(schema, child *apiextensions.JSONSchemaProps) CelSchemaContext {
+	r := c.childContext(schema)
+	r.initSchemaFields(child, false)
+	return r
+}
+
+// initSchemaFields converts the given schema and uses it to the Schema and DeclType fields of CelSchemaContext.
+// If the conversion fails Schema and DeclType are left nil since x-kubernetes-validations rules
+// are only allowed for valid structural schemas.
+func (c *CelSchemaContext) initSchemaFields(schema *apiextensions.JSONSchemaProps, isRoot bool) {
+	if schema != nil && len(schema.XValidations) > 0 {
+		// As a performance optimization, only perform the conversions to Schema and DeclType types if the schema has
+		// x-kubernetes-validations rules.
+		structural, declType, err := c.Converter(schema, isRoot)
```

Comment:
does this mean if I have x-kubernetes-validations at the root level and at a level nested 3 levels deep, I'll end up with two declTypes for those two levels?

Reply:
If the root level has a x-kubernetes-validations, the schema will be converted to a structural+declType and those types will be traversed while looking for deeper x-kubernetes-validations.  I've added a test that ensures that the number of conversions for cases like this is always 1.